components/common/src/templating/config.rs

```diff
@@ -588,34 +565,150 @@ fn is_toml_value_a_table(key: &str, table: &toml::value::Table) -> bool {
 }
 
 #[cfg(not(windows))]
-fn set_permissions<T: AsRef<Path>>(path: T, pkg: &Pkg) -> hcore::error::Result<()> {
+fn set_permissions<T: AsRef<Path>>(path: T, user: &str, group: &str) -> hcore::error::Result<()> {
     use crate::hcore::os::users;
     use crate::hcore::util::posix_perm;
 
     if users::can_run_services_as_svc_user() {
-        posix_perm::set_owner(path.as_ref(), &pkg.svc_user, &pkg.svc_group)?;
+        posix_perm::set_owner(path.as_ref(), &user, &group)?;
     }
-    posix_perm::set_permissions(path.as_ref(), CONFIG_PERMISSIONS)
+
+    let permissions = if path.as_ref().is_dir() {
+        CONFIG_DIR_PERMISSIONS
+    } else {
+        CONFIG_PERMISSIONS
+    };
+    posix_perm::set_permissions(path.as_ref(), permissions)
 }
 
 #[cfg(windows)]
-fn set_permissions<T: AsRef<Path>>(path: T, _pkg: &Pkg) -> hcore::error::Result<()> {
+fn set_permissions<T: AsRef<Path>>(path: T, _user: &str, _group: &str) -> hcore::error::Result<()> {
     use crate::hcore::util::win_perm;
 
     win_perm::harden_path(path.as_ref())
 }
 
+/// Entry point to recursively walk the configuration directory and subdirectories to
+/// construct the list of template files
+///
+/// `dir` should be a directory that exists.
+fn load_templates<T>(dir: T) -> Result<TemplateRenderer>
+where
+    T: AsRef<Path>,
+{
+    let mut template = TemplateRenderer::new();
+    if dir.as_ref().is_dir() {
+        load_templates_recurse(&dir, &PathBuf::new(), &mut template)?;
+    } else {
+        debug!(
+            "Configuration directory is not a directory: {:?}",
+            dir.as_ref()
+        );
+    }
+    Ok(template)
+}
+
+/// Helper function for `load_templates` to recurse into the directory structure.
+///
+/// `dir` should be a directory
+fn load_templates_recurse<T>(
+    dir: T,
+    context: &Path,
+    mut template: &mut TemplateRenderer,
+) -> Result<()>
+where
+    T: AsRef<Path>,
+{
+    if let Ok(entries) = std::fs::read_dir(dir) {
+        for entry in entries {
+            if let Ok(entry) = entry {
+                // Skip any entries in the template directory which aren't files or directories
+                match entry.file_type() {
+                    Ok(file_type) => {
```

Comment:
And even one more (plus this obviates the comment in L625 with a useful log message):
```rust
        match entry.file_type() {
            Ok(file_type) if file_type.is_file() => {
                ...
            }
            Ok(file_type) if file_type.is_dir() => {
                ...
            }
            Ok(file_type) => trace!("Skipping non file/directory entry: {:?}", file_type),
            Err(e) => debug!("Failed to get file metadata for {:?} : {}", entry, e),
        }
```